### ISSUE TITLE
Can't Install Canvas

### ISSUE THREAD

**Issue by BobsterLobster** [Automattic/node-canvas#1015]
I was redirected here because I was issuing an error at node-gyp and one person said to report it to the node-canvas repo. I have read the installation guide.

## Issue or Feature

Can't install canvas.

Error:
```
PS C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev> npm install --save canvas

> canvas@1.6.7 install C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas
> node-gyp rebuild


C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas>if not defined npm_config_node_gyp (node "C:\Users\Ali\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
TRACKER : error TRK0005: Failed to locate: "CL.exe". The system cannot find the file specified. [C:\Users\Ali\Desktop\C
oding\Discord\MTFMisAJ-Dev\node_modules\canvas\build\canvas.vcxproj]


gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Users\Ali\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:125:13)
gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Ali\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas
gyp ERR! node -v v8.6.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! canvas@1.6.7 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the canvas@1.6.7 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Ali\AppData\Roaming\npm-cache\_logs\2017-10-05T22_51_07_127Z-debug.log
PS C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev>
```

## Your Environment
* Canvas Version: I don't have it installed.
* Environment: Node 8.6.0 Windows 10

**Comment by jakedowns** [comment]
assuming you followed the wiki instructions here: 
https://github.com/Automattic/node-canvas/wiki/_pages
https://github.com/Automattic/node-canvas/wiki/Installation---Windows

i had to manually download the VS Community Installer from here:
https://www.visualstudio.com/downloads/?utm_source=mscom&utm_campaign=msdocs

then go to Modify > and make sure that Workloads "Desktop development with C++" was installed
then, finally CL.exe was installed to: 
`Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.11.25503\bin\HostX64\x86\cl.exe`

after that, `npm i canvas` finally worked. hope this helps

**Comment by MellyTheFireyMelly** [comment]
Alright I'll, try this when I get home

**Comment by MellyTheFireyMelly** [comment]
Okay so the `CL.exe` file appeared. But the error still happened. [Image](https://gyazo.com/3026ddf144cbd5681ab085d98223faf0.png)
```
PS C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev> npm i --save canvas

> canvas@1.6.7 install C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas
> node-gyp rebuild


C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas>if not defined npm_config_node_gyp (node "C:\Users\Ali\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
TRACKER : error TRK0005: Failed to locate: "CL.exe". The system cannot find the file specified. [C:\Users\Ali\Desktop\C
oding\Discord\MTFMisAJ-Dev\node_modules\canvas\build\canvas.vcxproj]


gyp ERR! build error
gyp ERR! stack Error: `C:\Program Files (x86)\MSBuild\14.0\bin\msbuild.exe` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Users\Ali\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:125:13)
gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Ali\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Ali\Desktop\Coding\Discord\MTFMisAJ-Dev\node_modules\canvas
gyp ERR! node -v v8.6.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
npm WARN discord.js@12.0.0-dev requires a peer of bufferutil@^3.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of erlpack@discordapp/erlpack but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of node-opus@^0.2.0 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of opusscript@^0.0.3 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of sodium@^2.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of libsodium-wrappers@^0.5.0 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js@12.0.0-dev requires a peer of uws@^8.14.0 but none is installed. You must install peer dependencies yourself.
npm WARN discord.js-commando@0.9.0 requires a peer of discord.js@^12.0.0 but none is installed. You must install peer dependencies yourself.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! canvas@1.6.7 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the canvas@1.6.7 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Ali\AppData\Roaming\npm-cache\_logs\2017-10-08T07_01_25_440Z-debug.log
```

I installed the `Desktop Develompent With C++`, help!

**Comment by medstripes** [comment]
I have the same issue in macOS also.

**Comment by kreftk** [comment]
@medstripes did you solve it?

**Comment by epistemex** [comment]
Get the same error here, but with canvas@next. GTK is installed, Python and msbuild. I have also tried to run node-gyp manually directly with a download from the git rep. using `node-gyp configure build` with the same error.

```text
> npm i canvas@next
gyp ERR! build error
gyp ERR! stack Error: `msbuild` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\
npm\node_modules\node-gyp\lib\build.js:258:23)
gyp ERR! stack     at emitTwo (events.js:126:13)
gyp ERR! stack     at ChildProcess.emit (events.js:214:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_proces
s.js:198:12)
gyp ERR! System Windows_NT 6.3.[--deleted--]
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodej
s\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\ [--deleted--] \node_modules\canvas
gyp ERR! node -v v8.9.3
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! canvas@2.0.0-alpha.9 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the canvas@2.0.0-alpha.9 install script.
npm ERR! This is probably not a problem with npm. There is likely additional log
ging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\[--deleted--]\AppData\Roaming\npm-cache\_logs\2018-02-15T11_35_27_18
6Z-debug.log
>
```

**Comment by ArthurMalet** [comment]
Same issue on Mac OS High Sierra.
Any ideas on how to fix this..?

`> canvas@1.6.11 install /Users/me/Dev/project/node_modules/canvas
> node-gyp rebuild

gyp ERR! configure error 
gyp ERR! stack Error: EACCES: permission denied, mkdir '/Users/me/Dev/project/node_modules/canvas/build'
gyp ERR! System Darwin 17.5.0
gyp ERR! command "/usr/local/Cellar/node/9.5.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/me/Dev/project/node_modules/canvas
gyp ERR! node -v v9.5.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
npm WARN socket@1.0.0 No description
npm WARN socket@1.0.0 No repository field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! canvas@1.6.11 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the canvas@1.6.11 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/me/.npm/_logs/2018-06-25T12_51_45_488Z-debug.log`

**Comment by micheleriva** [comment]
Same here.

```sh
 Node v8.9.4 micheleriva  λ ~> yarn add canvas
yarn add v1.9.4
warning package.json: No license field
warning No license field
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 📃  Building fresh packages...
error /Users/micheleriva/Desktop/projects/js/generative/node_modules/canvas: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: /Users/micheleriva/Desktop/projects/js/generative/node_modules/canvas
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@9.11.2 | darwin | x64
gyp info spawn /usr/bin/python
gyp info spawn args [ '/Users/micheleriva/.nvm/versions/node/v9.11.2/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/micheleriva/Desktop/projects/js/generative/node_modules/canvas/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/micheleriva/.nvm/versions/node/v9.11.2/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/micheleriva/.node-gyp/9.11.2/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/micheleriva/.node-gyp/9.11.2',
gyp info spawn args   '-Dnode_gyp_dir=/Users/micheleriva/.nvm/versions/node/v9.11.2/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/micheleriva/.node-gyp/9.11.2/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/micheleriva/Desktop/projects/js/generative/node_modules/canvas',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
Package fontconfig was not found in the pkg-config search path.
Perhaps you should add the directory containing `fontconfig.pc'
to the PKG_CONFIG_PATH environment variable
Package 'fontconfig', required by 'cairo', not found
gyp: Call to './util/has_lib.sh freetype' returned exit status 0 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/Users/micheleriva/.nvm/versions/node/v9.11.2/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:336:16)
gyp ERR! stack     at ChildProcess.emit (events.js:180:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:209:12)
gyp ERR! System Darwin 16.7.0
gyp ERR! command "/Users/micheleriva/.nvm/versions/node/v9.11.2/bin/node" "/Users/micheleriva/.nvm/versions/node/v9.11.2/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/micheleriva/Desktop/projects/js/generative/node_modules/canvas
gyp ERR! node -v v9.11.2
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**Comment by LinusU** [comment]
@micheleriva 

> Package fontconfig was not found in the pkg-config search path.
> Perhaps you should add the directory containing `fontconfig.pc'
> to the PKG_CONFIG_PATH environment variable
> Package 'fontconfig', required by 'cairo', not found

Have you installed `fontconfig`?

**Comment by micheleriva** [comment]
@LinusU yes, but I just figured out how to fix this error.

At first I was installing all the listed dependecies using brew: `brew install pkg-config cairo pango libpng jpeg giflib`.

Now I've tried using MacPorts (as written in the docs):

`port install pkgconfig cairo pango libpng jpeg giflib libsrvg`

this solved the issue!

**Comment by BobsterLobster** [comment]
I just installed with `npm i canvas@next`.

**Comment by epistemex** [comment]
For those with problem even after using the distributions including pre-compiled versions etc.,  this is what I had to do:

- First, I needed to remove anything Visual Studio completely (I've had several versions installed over the years)
- But (!) the VS' uninstaller won't do the job - I had to use [this special tool](https://github.com/Microsoft/VisualStudioUninstaller) to clean up most of the left-overs - only then would the build tools install correctly.
- Then I installed [this npm](https://github.com/felixrieseberg/windows-build-tools).
- I can't remember for sure, but I think I also had to install the [build tools distribution](https://blogs.msdn.microsoft.com/vcblog/2015/11/02/announcing-visual-c-build-tools-2015-standalone-c-tools-for-build-environments/) itself despite the previous step, but try first without.
- Also check that Python 2.7 exist and is available in PATH.
- Check your ENV setting for `VCTargetsPath` and that it points to for example `C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\v140`
- See canvas' own installation guides for GTK etc.
- (todo did I forget something...?)

And voila..
